### PR TITLE
Specification tip subsets in liquid handling operations

### PR DESCRIPTION
**Pull request by lhelleckes**
This PR adds functionality to not only specify one specific Tecan tip per aspirate or dispense, but rather lists of tips to be used.

For this, the `_pepare_aspirate_dispense_parameters` function was modified to accept integers between 1 and 8 (referring to the tips), `evotools.Tip` objects or iterables containing elements of the former two.

Tests were added respectively.

### CODE REVIEW

File: robotools/evotools/__init__.py [hunk 1]
```diff
@@ -94,25 +94,50 @@ def _prepate_aspirate_dispense_parameters(
...
-        elif tip == 8:
-            tip = Tip.T8
+        tip = _int_to_tip(tip)
+
     if not isinstance(tip, Tip):
-        raise ValueError(f'Invalid tip: {tip}')
+        try:
+            len(tip) > 0
+        except:
+            raise ValueError(f'tip must be an int between 1 and 8, Tip or Iterable, but was {tip}.')
```
Comment (michaelosthege) [inline robotools/evotools/__init__.py:126]: If you invert this if condition the else branch might be easier to read
```
# previous code dealt with int->Tip
if isinstance(tip, collections.abc.Iterable):
    # TODO: Iterate the elements to check their type & convert them to Tip
    tip = sum(set(tips))
elif not isinstance(tip, Tip):
    raise ValueError("Wrong tip type.")
```

File: robotools/tests.py [hunk 2]
```diff
@@ -247,66 +247,79 @@ def test_context(self):
...
         with self.assertRaises(ValueError):
-            evotools._prepate_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, rack_id=None)
+            evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=[1, 2.6])
+        evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=4)
+
+        _, _, _, _, tip, _, _, _, _ = evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=[evotools.Tip.T1, 4])
+        _, _, _, _, tip2, _, _, _, _ = evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=[1, 4])
+        _, _, _, _, tip3, _, _, _, _ = evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=[1, evotools.Tip.T4])
+        self.assertEqual(tip, tip2)
+        self.assertEqual(tip2, tip3)
```
Comment (michaelosthege) [inline robotools/tests.py:297]:
```suggestion
        _, _, _, _, tip, _, _, _, _ = evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=[evotools.Tip.T1, 4])
        assert tip == 17
        _, _, _, _, tip, _, _, _, _ = evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=[1, 4])
        assert tip == 17
        _, _, _, _, tip, _, _, _, _ = evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=[1, evotools.Tip.T4])
        assert tip == 17
```

File: robotools/tests.py [hunk 3]
```diff
@@ -247,66 +247,79 @@ def test_context(self):
...
-            evotools._prepate_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=None)
+            evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=None)
         with self.assertRaises(ValueError):
-            evotools._prepate_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=12)
-        evotools._prepate_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=4)
-        evotools._prepate_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=evotools.Tip.T5)
+            evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=12)
         with self.assertRaises(ValueError):
-            evotools._prepate_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, rack_id=None)
+            evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=[1, 2.6])
```
Comment (michaelosthege) [inline robotools/tests.py:290]: Please match the exception message. You can use pytest: `pytest.raises(ValueError, match="some part of the message")`

File: robotools/tests.py [hunk 4]
```diff
@@ -247,66 +247,79 @@ def test_context(self):
...
+        evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=4)
```
Comment (michaelosthege) [inline robotools/tests.py:291]:
```suggestion
        assert evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=4)[4] == 16
```

File: robotools/tests.py [hunk 5]
```diff
@@ -247,66 +247,79 @@ def test_context(self):
...
+
+        _, _, _, _, tip, _, _, _, _ = evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=[evotools.Tip.T1, 4])
+        _, _, _, _, tip2, _, _, _, _ = evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=[1, 4])
+        _, _, _, _, tip3, _, _, _, _ = evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=[1, evotools.Tip.T4])
+        self.assertEqual(tip, tip2)
+        self.assertEqual(tip2, tip3)
         with self.assertRaises(ValueError):
-            evotools._prepate_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, rack_id='invalid;rack')
+            evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=(evotools.Tip.T1, evotools.Tip.Any))
```
Comment (michaelosthege) [inline robotools/tests.py:299]: Also here please don't follow my bad example, but match the message with `pytest.raises(, match=...)`

File: robotools/tests.py [hunk 6]
```diff
@@ -247,66 +247,79 @@ def test_context(self):
...
+        
+        evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=evotools.Tip.Any)
+        evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=(evotools.Tip.T1, 4))
+        evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=evotools.Tip.T5)
```
Comment (michaelosthege) [inline robotools/tests.py:303]: Like above, these should assert something.

**Review by michaelosthege — CHANGES_REQUESTED** [review]
Two important things:
* According to https://stackoverflow.com/a/4668647/4473230 we should do `isinstance(tip, collections.abc.Iterable)`
* A `Tip` already is an `int`: `robotools.Tip.T3.__class__.__mro__` 👉 `(<enum 'Tip'>, <enum 'IntEnum'>, <class 'int'>, <enum 'Enum'>, <class 'object'>)` (The implementation is aware of this already, but just in case)

There's quite a bit of if/else branching that's hard to follow.
But with the above only one `if not isinstance(tip, (int, collections.abc.Iterable)): raise` check should be enough. Maybe a second one for the elements in the iterable.

File: robotools/tests.py (hunk 4 above)
Comment (lhelleckes) [reply robotools/tests.py:291]: The 4th tip is 8 ;)

File: robotools/tests.py (hunk 5 above)
Comment (lhelleckes) [reply robotools/tests.py:296]: Done

File: robotools/tests.py (hunk 6 above)
Comment (lhelleckes) [reply robotools/tests.py]: Thanks for the hint

File: robotools/evotools/__init__.py (hunk 1 above)
Comment (lhelleckes) [reply robotools/evotools/__init__.py]: Good point, thanks!

File: robotools/evotools/__init__.py [hunk 7]
```diff
@@ -94,25 +94,48 @@ def _prepate_aspirate_dispense_parameters(
...
+        tip = _int_to_tip(tip)
+
+    if isinstance(tip, collections.Iterable):
+        tips = []
+        for element in tip:
+            if isinstance(element, int) and not isinstance(element, Tip):
+                tips.append(_int_to_tip(element))
+            elif isinstance(element, Tip):
+                if element == -1:
+                    raise ValueError("When Iterables are used, the no Tip.Any objects are allowed. Use only Tip.Any instead.")
```
Comment (michaelosthege) [inline robotools/evotools/__init__.py:131]:
```suggestion
                    raise ValueError("When Iterables are used, no Tip.Any elements are allowed. Pass just one Tip.Any instead.")
```

File: robotools/tests.py [hunk 8]
```diff
@@ -247,66 +247,86 @@ def test_context(self):
...
+        _, _, _, _, tip, _, _, _, _ = evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=[evotools.Tip.T1, 4])
+        assert tip == 9
+        _, _, _, _, tip, _, _, _, _ = evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=[1, 4])
+        assert tip == 9
+        _, _, _, _, tip, _, _, _, _ = evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=[1, evotools.Tip.T4])
+        assert tip == 9
+        _, _, _, _, tip, _, _, _, _ = evotools._prepare_aspirate_dispense_parameters(rack_label='WaterTrough', position=1, volume=15, tip=evotools.Tip.Any)
+        assert tip == ''
+    
+        with pytest.raises(ValueError, match='no Tip.Any objects are allowed'):
```
Comment (michaelosthege) [inline robotools/tests.py:301]:
```suggestion
        with pytest.raises(ValueError, match='no Tip.Any elements are allowed'):
```

File: robotools/evotools/__init__.py [hunk 9]
```diff
@@ -35,7 +35,7 @@ class InvalidOperationError(Exception):
     pass
 
 
-def _prepate_aspirate_dispense_parameters(
+def _prepare_aspirate_dispense_parameters(
     rack_label:str, position:int, volume:float,
     liquid_class:str='',
     tip:typing.Union[Tip, int]=Tip.Any,
```
Comment (michaelosthege) [inline robotools/evotools/__init__.py:41]:
```suggestion
    tip:typing.Union[Tip, int, collections.abc.Iterable]=Tip.Any,
```

File: robotools/evotools/__init__.py [hunk 10]
```diff
@@ -94,25 +94,48 @@ def _prepate_aspirate_dispense_parameters(
...
-        elif tip == 7:
-            tip = Tip.T7
-        elif tip == 8:
-            tip = Tip.T8
-    if not isinstance(tip, Tip):
-        raise ValueError(f'Invalid tip: {tip}')
+        # User-specified integers from 1-8 need to be converted to Tecan logic
+        tip = _int_to_tip(tip)
+
+    if isinstance(tip, collections.Iterable):
```
Comment (michaelosthege) [inline robotools/evotools/__init__.py:124]:
```suggestion
    if isinstance(tip, collections.abc.Iterable):
```
Quoting [from stackoverflow](https://stackoverflow.com/a/4668647/4473230):
> Note: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.9 it will stop working.

**Review by michaelosthege — CHANGES_REQUESTED** [review]
One more grammar thing in an exception message, one type hint, and an important import detail..
Otherwise it's looking good.

**Review by michaelosthege — APPROVED** [review]
